core/src/test/scala/org/typelevel/paiges/Generators.scala

```diff
@@ -93,4 +94,18 @@ object Generators {
 
   implicit val cogenDoc: Cogen[Doc] =
     Cogen[Int].contramap((d: Doc) => d.hashCode)
+
+  implicit val shrinkDoc: Shrink[Doc] = {
+    import Doc._
+    Shrink {
+      case Union(a, b) =>
+        (shrink(a) map (Union(_, b))) append (shrink(b) map (Union(a, _)))
+      case Concat(a, b) =>
+        (shrink(a) map (Concat(_, b))) append (shrink(b) map (Concat(a, _)))
+      case Text(s) => shrink(s) map Text
```

Comment:
I think `shrink` may not be okay with `Text` because we have invariants. I think you want `shrink(s).map(text(_))`

Reply:
Yeah, where possible we should use our factory constructors to rebuild the shrunken asts.